server/balancer.go

```diff
@@ -214,11 +214,18 @@ func (cb *capacityBalancer) Balance(cluster *clusterInfo) (*balanceOperator, err
 		return nil, nil
 	}
 
-	leaderTransferOperator := newTransferLeaderOperator(oldLeader, newLeader, maxWaitCount)
+	// Third, select one region peer to do leader transfer.
+	newLeader := cb.selectNewLeaderPeer(cluster, followerPeers, newPeer)
+	if newLeader == nil {
+		log.Warn("new leader peer cannot be found to do balance")
+		return nil, nil
+	}
+
 	addPeerOperator := newAddPeerOperator(newPeer)
+	leaderTransferOperator := newTransferLeaderOperator(oldLeader, newLeader, maxWaitCount)
 	removePeerOperator := newRemovePeerOperator(oldLeader)
 
-	return newBalanceOperator(region, leaderTransferOperator, addPeerOperator, removePeerOperator), nil
+	return newBalanceOperator(region, addPeerOperator, leaderTransferOperator, removePeerOperator), nil
```

Comment:
If newLeader != newPeer, should we put leaderTransferOperator before addPeerOperator?